compiler/luci/import/src/CircleMetadataImporter.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CircleMetadataImporter.h"
+
+#include <set>
+#include <vector>
+
+namespace
+{
+
+uint32_t read_u32(const std::vector<uint8_t> &buffer, uint32_t idx)
+{
+  uint32_t val = 0;
+  val += (buffer.at(idx + 0) << 0 * 8);
+  val += (buffer.at(idx + 1) << 1 * 8);
+  val += (buffer.at(idx + 2) << 2 * 8);
+  val += (buffer.at(idx + 3) << 3 * 8);
+  return val;
+}
+
+} // namespace
+
+namespace
+{
+
+/**
+ * 'source_table' consists of following structure.
+ *  - [ entry_number : uint32_t ]
+ *  - [ id : uint32_t ][ length : uint32_t ][ data : 'length' Bytes ] * entry_number
+ *
+ * This will be decoded as std::map with following key and value.
+ *  - key : id of origin node
+ *  - value : name of origin node
+ *
+ * <Example>
+ *  - binary data : 0x01 00 00 00 00 00 00 00 0x04 00 00 00 0x6f 0x66 0x6d 00
+ *  - decoded data : {0 : "ofm"}
+ */
+const std::map<uint32_t, std::string>
+decoded_source_table(const std::vector<uint8_t> &source_table_data)
+{
+  std::map<uint32_t, std::string> source_id_name_map;
+  uint32_t idx = 0;
+
+  uint32_t entry_number = read_u32(source_table_data, idx);
+  idx += sizeof(uint32_t);
+
+  while (idx < source_table_data.size())
+  {
+    uint32_t id = read_u32(source_table_data, idx);
+    idx += sizeof(uint32_t);
+
+    uint32_t length = read_u32(source_table_data, idx);
+    idx += sizeof(uint32_t);
+
+    // The last character of name is '\0'.
+    // However, '\0' will be ignored because it is just for indicating end of name.
```

Comment:
FYI, https://en.wikipedia.org/wiki/Null-terminated_string

Reply:
If we use the data as `char *` or `char []`, we should use '\0' character to find the end of string.
However, as we use the data as `std::string`, we do not need to insert '\0' to the string.

If the comment is confused, I will add additional explanation to it :)